### ISSUE TITLE
How To Identify SSL Handshake errors 

### ISSUE THREAD

**Issue by sratnesh** [openssl/openssl#12958]
Hi , 

I want to log errors which occurs during SSL handshake 

One example is : 

when I set the min proto version to TLS1_3_VERSION
SSL_CTX_set_min_proto_version(m_sslCtx, TLS1_3_VERSION);

But server only supports TLS1_2_VERSION

So in this case when i do handShakeStatus = SSL_do_handshake(m_ssl);

the error i get using  int err = SSL_get_error(m_ssl, handShakeStatus);

is 5 which is 

> # define SSL_ERROR_SYSCALL    5/* look at error stack/return
>                                            * value/errno */

and in case of The failure in finding a matching cipher suite during the TLS handshake  also I am getting same error code 

Now I want to log the exact error to error log in my application 

So how to get exact error from SSL_ERROR_SYSCALL  error which i get in all cases 

How i can identify exact error 

and SSL library says /* look at error stack/return
                                   * value/errno */
But there is no clarity how to look and how exactly ?

Need urgent help on this 

Thanks
Ratnesh

**Comment by sratnesh** [comment]
Please guide me how to log errors and which API to use for below Errors / Issues 

1)  mis match in cipher suite during the TLS handshake error
2) use of  older version than TLS 1.0 or SSL 3.1 error
3) termination of the session error
4)  termination of a connection due to a missed TLS session renegotiation error
5) Exceeding the maximum size of a certificate during a TLS handshake error
6) The connection termination due to the lack of a certificate of either side error
7) The failure of finding a matching CA certificate during a TLS handshake error

Please help in in getting exact error log / error code for above cases

**Comment by mattcaswell** [comment]
Please look at the `ERR_*` function calls in the man pages. For example `ERR_print_errors` can be used to dump the error stack to a BIO.

**Comment by sratnesh** [comment]
@mattcaswell  Do you mean If I use below call then I will get all errors ( from point 1 to point 7) ?


string getOpenSSLError()
{
	BIO *bio = BIO_new(BIO_s_mem());
	ERR_print_errors(bio);
	char *buf;
	size_t len = BIO_get_mem_data(bio, &buf);
	string ret(buf, len);
	BIO_free(bio);
	return ret;
}

**Comment by sratnesh** [comment]
@mattcaswell 
After making this call I got error string as 

"16260:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl\\record\\rec_layer_s3.c:1544:SSL alert number 40\n"

So How I can Interpret this error ?

Exactly How I can categorize what is the issue ?

Where I can see all Alert number definitions ?

**Comment by sratnesh** [comment]
@mattcaswell 

Suppose from above error I want to classify error type as "mis match in cipher suite during the TLS handshake error"

Then exactly what I should do ?

**Comment by mattcaswell** [comment]
> Do you mean If I use below call then I will get all errors ( from point 1 to point 7) ?

Yes. Well that's one way to do it. There are quite a few different ERR_* functions that enable you to examine each individual error on the error stack, examine its reason code, get a human readable string, etc. Please take a look at the various ERR_* man pages:

https://www.openssl.org/docs/man1.1.1/man3/

> "16260:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl\record\rec_layer_s3.c:1544:SSL alert number 40\n"

This means you've received a handshake failure alert from the peer. So the peer is telling you it is unable to proceed with the handshake for some reason. There is no more detailed information from the peer in this case. To find out the reason you'd have to examine the peer's logs.

> Where I can see all Alert number definitions ?

All the various SSL/TLS specific error reason codes are listed in openssl/sslerr.h (included automatically from ssl.h). The ones specific to alerts all have `_ALERT_` in the name. You can retrieve the earliest error code using ERR_get_error() and you can extract the reason code from an error code using `ERR_GET_REASON`.

You should check the various RFCs for detailed descriptions of what the alerts actually mean.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
